### ISSUE TITLE
rsync transfer process hangs indefinitely

### ISSUE THREAD

**Issue by agis** [WayneD/rsync#87]
Hello,

First of all, thanks for rsync and apologies for not trying with the latest (3.2.3), but it's hard to upgrade since this is a production service that uses the Debian package. That said, I can do so if you believe it is needed.

Our use case: we have a single server that runs rsyncd and serves a large dataset, currently at _422GB_ (and is increasing over time). Then we have a bunch of other servers (currently 4 but also increasing over time) that all download _at the same time_, this dataset from the rsyncd host, using the following command:

```
rsync -aPH --chown=mysql:mysql --delete --inplace --bwlimit=40m rsync://${RSYNC_SERVER}/ezra/ ${MYSQL_WORK_DIR}
```

We notice that sometimes the rsync transfers hang. Eventually we applied a `--bwlimit`, which merely reduced the frequency at which this happens. This issue we have is back from 2017, when we were using rsync 3.1.2. So recently we noticed that rsync 3.2.1 included a relevant commit (01b9bbb0f94) so we upgraded to it - it also reduced the frequency (I think) but the issue still happens occasionally.

Another observation is that the issue seems to be getting worse with the increase in the dataset and in the number of receiving hosts.

Below I'm attaching information which I believe may help to debug this (let me know if anything else is needed).

Thanks in advance!

rsync logs
---------------

Sending end (rsyncd):

```
2020/08/29 01:00:01 [23392] rsync on ezra/ from ezra4.foobar.com ([IPv6 REDACTED])[IPv6 REDACTED])
2020/08/29 01:00:01 [23392] building file list
```

Process tree
---------------

Receiving end:

```
root     26091  0.2  0.0  41796  5060 ?        S    Aug29   8:36 \_ rsync -aPH --chown=mysql:mysql --delete --inplace --bwlimit=40m rsync://strife.foobar.com/ezra/ /srv/ezra/data/mysql
root     26092  0.3  0.0  41828  3560 ?        S    Aug29  11:41     \_ rsync -aPH --chown=mysql:mysql --delete --inplace --bwlimit=40m rsync://strife.foobar.com/ezra/ /srv/ezra/data/mysql
```

Sending end (including daemon process):

```
root     26458  0.0  0.0  26680  1308 ?        Ss   Jul22   0:00 /usr/bin/rsync --daemon --no-detach
mysql    23392  2.8  0.0  54596 15412 ?        S    Aug29  90:23  \_ /usr/bin/rsync --daemon --no-detach
```

Send/Receive queues
---------------

Receiving end:

```
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp6       0 529823 ezra4.foobar.com:32840  strife.foobar:rsync ESTABLISHED 26091/rsync
```

Sending end (rsyncd):

```
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp6       0      0 strife.foobar:rsync ezra4.foobar.com:32840  ESTABLISHED 23392/rsync
```

System call trace
==================

Receiving end (pid=26091):

```
select(5, [4], [3], [4], {tv_sec=59, tv_usec=166979}) = 0 (Timeout)
select(5, [4], [3], [4], {tv_sec=60, tv_usec=0}) = 0 (Timeout)
select(5, [4], [3], [4], {tv_sec=60, tv_usec=0}) = 0 (Timeout)
select(5, [4], [3], [4], {tv_sec=60, tv_usec=0}) = 0 (Timeout)
select(5, [4], [3], [4], {tv_sec=60, tv_usec=0}) = 0 (Timeout)
select(5, [4], [3], [4], {tv_sec=60, tv_usec=0}) = 0 (Timeout)
```

Receiving end (pid=26092):

```
select(4, [3], [], [3], {tv_sec=23, tv_usec=904884}) = 0 (Timeout)
select(4, [3], [], [3], {tv_sec=60, tv_usec=0}) = 0 (Timeout)
select(4, [3], [], [3], {tv_sec=60, tv_usec=0}) = 0 (Timeout)
select(4, [3], [], [3], {tv_sec=60, tv_usec=0}) = 0 (Timeout)
select(4, [3], [], [3], {tv_sec=60, tv_usec=0}) = 0 (Timeout)
select(4, [3], [], [3], {tv_sec=60, tv_usec=0}) = 0 (Timeout)
select(4, [3], [], [3], {tv_sec=60, tv_usec=0}) = 0 (Timeout)
select(4, [3], [], [3], {tv_sec=60, tv_usec=0}) = 0 (Timeout)
```

Sending end (pid=23392):

```
select(7, [6], [], [6], {tv_sec=47, tv_usec=71793}) = 0 (Timeout)
select(7, [6], [], [6], {tv_sec=60, tv_usec=0}) = 0 (Timeout)
select(7, [6], [], [6], {tv_sec=60, tv_usec=0}) = 0 (Timeout)
select(7, [6], [], [6], {tv_sec=60, tv_usec=0}) = 0 (Timeout)
select(7, [6], [], [6], {tv_sec=60, tv_usec=0}) = 0 (Timeout)
```

Network/TCP settings
------------------------------

The following are custom TCP settings tuned in an attempt to fix the issue.
Apparently, they didn't help.

Receiving end:

```
# settings for the socket receive buffer
net.core.rmem_default = 524288
net.core.rmem_max = 4194304

# settings for the socket send buffer
net.core.wmem_default = 524288

# tcp memory settings
net.ipv4.tcp_rmem = 100420 524288 4194304
net.ipv4.tcp_wmem = 100420 524288 4194304
```

System information
---------------

Receiving end:

- Debian GNU/Linux 9.6 (stretch)
- Linux ezra4 4.9.0-8-amd64 #1 SMP Debian 4.9.130-2 (2018-10-27) x86_64 GNU/Linux
- rsync  version 3.2.1  protocol version 31

Sending end:

- Debian GNU/Linux 9.5 (stretch)
- Linux strife 4.9.0-8-amd64 #1 SMP Debian 4.9.110-3+deb9u6 (2018-10-08) x86_64 GNU/Linux
- rsync  version 3.2.1  protocol version 31

rsyncd configuration
----------------

```
[ezra]
path = /srv/snapshots/ezra
lock file = /run/rsync/ezra.lock
uid = mysql
gid = mysql
comment = ezra MySQL data
max connections = 4
read only = true
hosts allow =  [IPv4 REDACTED] [IPv6 REDACTED]
exclude from = /srv/snapshots/ezra/.exclude
log file = /var/log/rsyncd.log
```

**Comment by edo1** [comment]
It looks like a network issue. I guess `--timeout=30` will help. rsync will fail, but you can make another try in a script, something like
```
TRY=0
until rsync --timeout=30 bla-bla-bla; do
    TRY=$((TRY+1))
    if [ $TRY -ge 5 ]; then
        echo something goes wrong, exiting >&2
        exit 1
    fi
done
```

**Comment by agis** [comment]
Assuming it's a transient network issue (a hiccup), isn't rsync expected to pick up where it left off after the network is back up again?

**Comment by WayneD** [comment]
Given that the one side has a huge send queue that isn't getting to the other side, and all the processes are waiting for data, it does look like your tcp connection has failed. You might try the `-4` option and see if tcp4 works better than tcp6 on your systems.

**Comment by realtalk69** [comment]
How to rectify

**Comment by apoikos** [comment]
For posterity, this turned out to be a kernel bug in the TCP receiver fast path, fixed in 5.10 (and a series of stable kernels). See https://lore.kernel.org/netdev/87eelz4abk.fsf@marvin.dmesg.gr/T/#u for details.

**Comment by Frankkkkk** [comment]
Thank you for the follow up @apoikos ! I had the same problem as your (v6 too) and your debug is really interesting !

If you come one day to Geneva, I'm paying you a beer.
Cheers

**Comment by apoikos** [comment]
@Frankkkkk glad I helped! We've been running a fixed kernel for 3 months now and have had zero hangs, so I'm pretty confident about the fix.

And of course I would never turn down an offer for a beer!

**Comment by lijxug** [comment]
> Thank you for the follow up @apoikos ! I had the same problem as you (v6 too) and your debug is really interesting !
> 
> If you come one day to Geneva, I'm paying you a beer. Cheers

Sorry I have to ask. How do I check if my machine have installed this fix? If not, how do I install it? 
I've encounter a similar problem and it's driving me crazy. Thanks a lot.

**Comment by jfragoulis** [comment]
@lijxug this was [merged at version 5.10-rc1 of the linux kernel](https://github.com/torvalds/linux/commit/18ded910b589839e38a51623a179837ab4cc3789). So, if you have a kernel version greater than that (which you probably have), you have the fix.
